File: apps/mobile/ios/Sources/FlagshipCore/ToastCenter.swift
import Foundation
import Observation

/// App-wide toast queue. Views publish `info` / `success` / `warning` /
/// `error` messages through here; the `Toaster` overlay at the root of
/// ContentView renders the top-most one as a transient banner.
///
/// Toasts are de-duplicated by `id` so rapid identical publishes don't
/// stack. Auto-dismiss after `duration`; tap dismisses immediately.
@Observable
@MainActor
public final class ToastCenter {
    public private(set) var queue: [Toast] = []

    public init() {}

    public func info(_ message: String, duration: TimeInterval = 3) {
        publish(Toast(kind: .info, message: message, duration: duration))
    }
    public func success(_ message: String, duration: TimeInterval = 2.5) {
        publish(Toast(kind: .success, message: message, duration: duration))
    }
    public func warning(_ message: String, duration: TimeInterval = 4) {
        publish(Toast(kind: .warning, message: message, duration: duration))
    }
    public func error(_ message: String, duration: TimeInterval = 5) {
        publish(Toast(kind: .error, message: message, duration: duration))
    }

    public func dismiss(_ id: UUID) {
        queue.removeAll { $0.id == id }
    }

    private func publish(_ toast: Toast) {
        if queue.contains(where: { $0.kind == toast.kind && $0.message == toast.message }) {
            return
        }
        queue.append(toast)
        let id = toast.id
        Task { [weak self] in
            try? await Task.sleep(nanoseconds: UInt64(toast.duration * 1_000_000_000))
            await MainActor.run { self?.dismiss(id) }
        }
    }
}

public struct Toast: Identifiable, Equatable, Sendable {
    public enum Kind: Sendable { case info, success, warning, error }
    public let id: UUID
    public let kind: Kind
    public let message: String
    public let duration: TimeInterval
    public init(id: UUID = UUID(), kind: Kind, message: String, duration: TimeInterval) {
        self.id = id
        self.kind = kind
        self.message = message
        self.duration = duration
    }
}

File: apps/mobile/android/app/src/main/java/com/flagship/core/ToastCenter.kt
// Kotlin mirror of FlagshipCore/ToastCenter.swift.
//
// App-wide toast queue. Views publish info/success/warning/error
// messages; a Toaster overlay at the root of the scaffold renders the
// top-most one. Toasts are de-duplicated by (kind, message) so rapid
// identical publishes don't stack.

package com.flagship.core

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.UUID
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds

class ToastCenter(
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default),
) {
    private val _queue = MutableStateFlow<List<Toast>>(emptyList())
    val queue: StateFlow<List<Toast>> = _queue.asStateFlow()

    fun info(message: String, duration: Duration = 3.seconds) =
        publish(Toast(kind = Toast.Kind.INFO, message = message, duration = duration))
    fun success(message: String, duration: Duration = 2500.milliseconds) =
        publish(Toast(kind = Toast.Kind.SUCCESS, message = message, duration = duration))
    fun warning(message: String, duration: Duration = 4.seconds) =
        publish(Toast(kind = Toast.Kind.WARNING, message = message, duration = duration))
    fun error(message: String, duration: Duration = 5.seconds) =
        publish(Toast(kind = Toast.Kind.ERROR, message = message, duration = duration))

    fun dismiss(id: String) {
        _queue.value = _queue.value.filterNot { it.id == id }
    }

    private fun publish(toast: Toast) {
        if (_queue.value.any { it.kind == toast.kind && it.message == toast.message }) return
        _queue.value = _queue.value + toast
        scope.launch {
            delay(toast.duration)
            dismiss(toast.id)
        }
    }
}

data class Toast(
    val id: String = UUID.randomUUID().toString(),
    val kind: Kind,
    val message: String,
    val duration: Duration,
) {
    enum class Kind { INFO, SUCCESS, WARNING, ERROR }
}

